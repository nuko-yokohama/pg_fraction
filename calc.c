/*
 * calc.c
 */

#include <stdio.h>
#include <stdlib.h>
#include "calc.h"

/*
 * reduce function.
 * return 1:reduced, 0:not reduced
 * int* numerator(in-out parameter)
 * int* denominator(in-out parameter)
 */
int reduce_fraction(int *n, int *d) {
    int loop_end;
    int r = 0; /* was not reduce */
    int i; /* loop counter */

    if ( *n == *d ) {
        *n = 1;
        *d = 1;
        return 1;
    }

    if ( *n == 0 ) {
        *n = 0;
        *d = 1;
        return 1;
    }

    if ( abs(*n) > abs(*d) ) {
        loop_end = abs(*d);
    } else {
        loop_end = abs(*n);
    }

    i = 2;
    while (i <= loop_end) {
        if ( (*n % i) == 0 && (*d % i) == 0) {
            /* Both the numerator and denominator were divisible */
            *n = *n / i;
            *d = *d / i;
            r = 1; /* was redueced */
        } else {
            i++; 
        }
    }

    /* convert minus sign */
    if ( *n < 0 && *d < 0 ) {
        *n = abs(*n);
        *d = abs(*d);
    }

    return r;
}

/*
 * Calculation Functions
 */

/*
 * Add Functoin
 */
int add_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od) {
    int t1n, t2n, td;

    t1n = n1 * d2;
    t2n = n2 * d1;
    td = d1 * d2;

    *on = t1n + t2n;
    *od = td;

    reduce_fraction(on, od);
    return 0;
}

/*
 * Sub Functoin
 */
int sub_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od) {
    int t1n, t2n, td;

    t1n = n1 * d2;
    t2n = n2 * d1;
    td = d1 * d2;

    *on = t1n - t2n;
    *od = td;

    reduce_fraction(on, od);
    return 0;
}

/*
 * Mul Functoin
 */
int mul_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od) {
    *on = n1 * n2;
    *od = d1 * d2;

    reduce_fraction(on, od);
    return 0;
}

/*
 * Div Functoin
 */
int div_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od) {
    *on = n1 * d2;
    *od = d1 * n2;

    reduce_fraction(on, od);
    return 0;
}


#ifdef REDUCE
int
main(int argc, char** argv) {
   int r;
   int n;
   int d;

   if (argc != 3) {
     printf("reduce num, demo\n");
     return 0;
   }

   n = atoi(argv[1]);
   d = atoi(argv[2]);
   
   printf("before r=%d, n=%d,d=%d\n", r, n, d);
   r = reduce_fraction( &n, &d );
   printf("after  r=%d, n=%d,d=%d\n", r, n, d);
}
#endif

#ifdef CALC
int
main(int argc, char** argv) {
   int r;
   int n1;
   int d1;
   int n2;
   int d2;
   int on;
   int od;
   double f1, f2;

   if (argc != 5) {
     printf("calc num1 demo1 num2 demo2\n");
     return 0;
   }

   n1 = atoi(argv[1]);
   d1 = atoi(argv[2]);
   n2 = atoi(argv[3]);
   d2 = atoi(argv[4]);
   
   f1 = (double) n1 / (double) d1;
   printf("n1=%d,d1=%d,f1=%f\n", n1, d1, f1);
   f2 = (double) n2 / (double) d2;
   printf("n2=%d,d2=%d,f2=%f\n", n2, d2, f2);

   r = add_fraction_internal( n1, d1, n2, d2, &on, &od );
   printf("add_fraction n1=%d,d1=%d,n2=%d,d2=%d,on=%d,od=%d\n", n1, d1, n2, d2,on,od);
   r = sub_fraction_internal( n1, d1, n2, d2, &on, &od );
   printf("sub_fraction n1=%d,d1=%d,n2=%d,d2=%d,on=%d,od=%d\n", n1, d1, n2, d2,on,od);
   r = mul_fraction_internal( n1, d1, n2, d2, &on, &od );
   printf("mul_fraction n1=%d,d1=%d,n2=%d,d2=%d,on=%d,od=%d\n", n1, d1, n2, d2,on,od);
   r = div_fraction_internal( n1, d1, n2, d2, &on, &od );
   printf("div_fraction n1=%d,d1=%d,n2=%d,d2=%d,on=%d,od=%d\n", n1, d1, n2, d2,on,od);

}
#endif
