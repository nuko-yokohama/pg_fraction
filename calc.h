/*
 * calc.h
 * define calc-utility functions.
 *
 */

extern int reduce_fraction(int *n, int *d);
extern int add_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od);
extern int sub_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od);
extern int mul_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od);
extern int div_fraction_internal(int n1, int d1, int n2, int d2, int *on, int *od);

