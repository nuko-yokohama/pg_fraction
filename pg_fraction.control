# pg_fraction extension
comment = 'fraction data type'
default_version = '1.0'
module_pathname = '$libdir/pg_fraction'
relocatable = true
