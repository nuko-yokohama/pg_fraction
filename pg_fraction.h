/*
 * pg_fraction.h
 *
 * define fraction type 
 *
 */

/*
 * Internal representation of fraction type
 */
typedef struct st_fraction {
    int numerator;
    int denominator;
    double value;
} fraction;

