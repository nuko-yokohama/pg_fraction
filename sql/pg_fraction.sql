CREATE EXTENSION IF NOT EXISTS pg_fraction;

CREATE TABLE test (id int, data fraction);
INSERT INTO test VALUES (1, '2/3'),(2, '3/7'),(3,'4/20');
TABLE test;

SELECT id, data, data + '2/3' AS add, data - '2/3' AS sub FROM test;
SELECT id, data, data * '2/3' AS mul, data / '2/3' AS div FROM test;

SELECT max(data), min(data) FROM test;

SELECT id, data FROM test WHERE data = '5/25';
SELECT id, data FROM test WHERE data <> '5/25';
SELECT id, data FROM test WHERE data < '5/25';
SELECT id, data FROM test WHERE data > '5/25';
SELECT id, data FROM test WHERE data <= '5/25';
SELECT id, data FROM test WHERE data >= '5/25';

-- order by
SELECT id, data, get_value(data) FROM test ORDER BY data;

-- cast
SELECT '1/3'::fraction::float8;
SELECT 3::fraction;

