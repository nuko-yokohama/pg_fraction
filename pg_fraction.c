/*
 * pg_fraction.c
 * fraction type postgresql functions
 *
 */

#include <string.h>

#include "postgres.h"
#include "fmgr.h"
#include "utils/geo_decls.h"

#include "calc.h"

/* Version-1 protocol */
#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

/*
 * Internal representation of fraction type
 */
typedef struct ST_Fraction {
    int numerator;
    int denominator;
    double value;
} Fraction;

/*
 * fraction_in()
 * input fraction type 
 */
PG_FUNCTION_INFO_V1(fraction_in);
Datum
fraction_in(PG_FUNCTION_ARGS)
{
    char       *str = PG_GETARG_CSTRING(0);
    int         n, d;
    Fraction   *result;

    /* TODO: more smart parse */

    if (sscanf(str, "%d/%d", &n, &d) != 2)
        ereport(ERROR,
                (errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
                 errmsg("pg_fraction:invalid input syntax: \"%s\"",
                        str)));
    /* range check */
    if (d == 0) {
        ereport(ERROR,
                (errcode(ERRCODE_DIVISION_BY_ZERO),
                 errmsg("pg_fraction:division by zero: \"%s\"",
                        str)));
    }

    result = (Fraction* ) palloc(sizeof(Fraction));
    reduce_fraction(&n, &d);

    if ( abs(n) > 99999 || abs(d) > 99999 ) {
        ereport(ERROR,
                (errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
                 errmsg("pg_fraction:: overflow error (max=99999) \"%s\"",
                        str)));
    }

    result->numerator = n;
    result->denominator = d;
    result->value = (double) n / (double) d;
    PG_RETURN_POINTER(result);
}


/*
 * fraction_out()
 * output fraction type 
 */
PG_FUNCTION_INFO_V1(fraction_out);
Datum
fraction_out(PG_FUNCTION_ARGS)
{
    Fraction   *fraction= (Fraction*) PG_GETARG_POINTER(0);
    char       *result;

    result = psprintf("%d/%d", fraction->numerator, fraction->denominator);
    PG_RETURN_CSTRING(result);
}

/*
 * get_value(Fractoin)
 * get double value
 */
PG_FUNCTION_INFO_V1(get_value);
Datum get_value(PG_FUNCTION_ARGS);

Datum get_value(PG_FUNCTION_ARGS) {
    Fraction *f1;
    f1 = (Fraction*) PG_GETARG_POINTER(0);
    PG_RETURN_FLOAT8(f1->value);
}

/*
 * Four arithmetic functions 
 */

PG_FUNCTION_INFO_V1(fraction_add);
Datum fraction_add(PG_FUNCTION_ARGS);

Datum fraction_add(PG_FUNCTION_ARGS) {
    Fraction *f1;
    Fraction *f2;
    Fraction *result;

    int n, d;

    f1 = (Fraction*) PG_GETARG_POINTER(0);
    f2 = (Fraction*) PG_GETARG_POINTER(1);

    add_fraction_internal(
        f1->numerator,
        f1->denominator,
        f2->numerator,
        f2->denominator,
        &n,
        &d);

    result = (Fraction*) palloc(sizeof(Fraction));
    result->numerator = n;
    result->denominator = d;
    result->value = (double) n / (double) d;

    PG_RETURN_POINTER(result);
}

PG_FUNCTION_INFO_V1(fraction_sub);
Datum fraction_sub(PG_FUNCTION_ARGS);

Datum fraction_sub(PG_FUNCTION_ARGS) {
    Fraction *f1;
    Fraction *f2;
    Fraction *result;

    int n, d;

    f1 = (Fraction*) PG_GETARG_POINTER(0);
    f2 = (Fraction*) PG_GETARG_POINTER(1);

    sub_fraction_internal(
        f1->numerator,
        f1->denominator,
        f2->numerator,
        f2->denominator,
        &n,
        &d);

    result = (Fraction*) palloc(sizeof(Fraction));
    result->numerator = n;
    result->denominator = d;
    result->value = (double) n / (double) d;

    PG_RETURN_POINTER(result);
}

PG_FUNCTION_INFO_V1(fraction_mul);
Datum fraction_mul(PG_FUNCTION_ARGS);

Datum fraction_mul(PG_FUNCTION_ARGS) {
    Fraction *f1;
    Fraction *f2;
    Fraction *result;

    int n, d;

    f1 = (Fraction*) PG_GETARG_POINTER(0);
    f2 = (Fraction*) PG_GETARG_POINTER(1);

    mul_fraction_internal(
        f1->numerator,
        f1->denominator,
        f2->numerator,
        f2->denominator,
        &n,
        &d);

    result = (Fraction*) palloc(sizeof(Fraction));
    result->numerator = n;
    result->denominator = d;
    result->value = (double) n / (double) d;

    PG_RETURN_POINTER(result);
}

PG_FUNCTION_INFO_V1(fraction_div);
Datum fraction_div(PG_FUNCTION_ARGS);

Datum fraction_div(PG_FUNCTION_ARGS) {
    Fraction *f1;
    Fraction *f2;
    Fraction *result;

    int n, d;

    f1 = (Fraction*) PG_GETARG_POINTER(0);
    f2 = (Fraction*) PG_GETARG_POINTER(1);

    div_fraction_internal(
        f1->numerator,
        f1->denominator,
        f2->numerator,
        f2->denominator,
        &n,
        &d);

    result = (Fraction*) palloc(sizeof(Fraction));
    result->numerator = n;
    result->denominator = d;
    result->value = (double) n / (double) d;

    PG_RETURN_POINTER(result);
}

/*
 * aggregate functions.
 */

/*
 * max
 */
PG_FUNCTION_INFO_V1(fraction_max);
Datum fraction_max(PG_FUNCTION_ARGS);

Datum fraction_max(PG_FUNCTION_ARGS) {
    Fraction* f_l = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f_r = (Fraction*) PG_GETARG_POINTER(1);


    if (f_l->value > f_r->value ) {
        PG_RETURN_POINTER(f_l);
    } else {
        PG_RETURN_POINTER(f_r);
    }
}

/*
 * max
 */
PG_FUNCTION_INFO_V1(fraction_min);
Datum fraction_min(PG_FUNCTION_ARGS);

Datum fraction_min(PG_FUNCTION_ARGS) {
    Fraction* f_l = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f_r = (Fraction*) PG_GETARG_POINTER(1);


    if (f_l->value < f_r->value ) {
        PG_RETURN_POINTER(f_l);
    } else {
        PG_RETURN_POINTER(f_r);
    }
}

/*
 * comparision functoins.
 */
static int
fraction_cmp_internal(double a, double b)
{
    if (a < b)
        return -1;
    if (a > b)
        return 1;
    return 0;
}

PG_FUNCTION_INFO_V1(fraction_eq);
Datum fraction_eq(PG_FUNCTION_ARGS);

Datum
fraction_eq(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) == 0);
}

PG_FUNCTION_INFO_V1(fraction_ne);
Datum fraction_ne(PG_FUNCTION_ARGS);

Datum
fraction_ne(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) != 0);
}

PG_FUNCTION_INFO_V1(fraction_lt);
Datum fraction_lt(PG_FUNCTION_ARGS);

Datum
fraction_lt(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) < 0);
}

PG_FUNCTION_INFO_V1(fraction_gt);
Datum fraction_gt(PG_FUNCTION_ARGS);

Datum
fraction_gt(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) > 0);
}

PG_FUNCTION_INFO_V1(fraction_le);
Datum fraction_le(PG_FUNCTION_ARGS);

Datum
fraction_le(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) <= 0);
}

PG_FUNCTION_INFO_V1(fraction_ge);
Datum fraction_ge(PG_FUNCTION_ARGS);

Datum
fraction_ge(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_BOOL(fraction_cmp_internal( f1_v, f2_v) >= 0);
}

PG_FUNCTION_INFO_V1(fraction_cmp);
Datum fraction_eq(PG_FUNCTION_ARGS);

/*
 * btree index support function.
 */
Datum
fraction_cmp(PG_FUNCTION_ARGS)
{
    Fraction* f1 = (Fraction*) PG_GETARG_POINTER(0);
    Fraction* f2 = (Fraction*) PG_GETARG_POINTER(1);
    double f1_v = f1->value;
    double f2_v = f2->value;

    PG_RETURN_INT32(fraction_cmp_internal( f1_v, f2_v));
}

/*
 * cast functions.
 */
PG_FUNCTION_INFO_V1(fraction_to_float8);
Datum fraction_to_float8(PG_FUNCTION_ARGS);

Datum
fraction_to_float8(PG_FUNCTION_ARGS)
{
    Fraction* a = (Fraction*) PG_GETARG_POINTER(0);

    PG_RETURN_FLOAT8( a->value );
}

PG_FUNCTION_INFO_V1(int32_to_fraction);
Datum int32_to_fraction(PG_FUNCTION_ARGS);

Datum
int32_to_fraction(PG_FUNCTION_ARGS)
{
    Fraction* f;
    int32 a =  PG_GETARG_INT32(0);
    f = (Fraction*) palloc(sizeof(Fraction));

    f->numerator = a;
    f->denominator = 1;
    f->value = (float8) a;

    PG_RETURN_POINTER(f);
}

