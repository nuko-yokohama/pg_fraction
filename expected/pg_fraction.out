CREATE EXTENSION IF NOT EXISTS pg_fraction;
CREATE TABLE test (id int, data fraction);
INSERT INTO test VALUES (1, '2/3'),(2, '3/7'),(3,'4/20');
TABLE test;
 id | data 
----+------
  1 | 2/3
  2 | 3/7
  3 | 1/5
(3 rows)

SELECT id, data, data + '2/3' AS add, data - '2/3' AS sub FROM test;
 id | data |  add  |  sub  
----+------+-------+-------
  1 | 2/3  | 4/3   | 0/1
  2 | 3/7  | 23/21 | -5/21
  3 | 1/5  | 13/15 | -7/15
(3 rows)

SELECT id, data, data * '2/3' AS mul, data / '2/3' AS div FROM test;
 id | data | mul  | div  
----+------+------+------
  1 | 2/3  | 4/9  | 1/1
  2 | 3/7  | 2/7  | 9/14
  3 | 1/5  | 2/15 | 3/10
(3 rows)

SELECT max(data), min(data) FROM test;
 max | min 
-----+-----
 2/3 | 1/5
(1 row)

SELECT id, data FROM test WHERE data = '5/25';
 id | data 
----+------
  3 | 1/5
(1 row)

SELECT id, data FROM test WHERE data <> '5/25';
 id | data 
----+------
  1 | 2/3
  2 | 3/7
(2 rows)

SELECT id, data FROM test WHERE data < '5/25';
 id | data 
----+------
(0 rows)

SELECT id, data FROM test WHERE data > '5/25';
 id | data 
----+------
  1 | 2/3
  2 | 3/7
(2 rows)

SELECT id, data FROM test WHERE data <= '5/25';
 id | data 
----+------
  3 | 1/5
(1 row)

SELECT id, data FROM test WHERE data >= '5/25';
 id | data 
----+------
  1 | 2/3
  2 | 3/7
  3 | 1/5
(3 rows)

-- order by
SELECT id, data, get_value(data) FROM test ORDER BY data;
 id | data |     get_value     
----+------+-------------------
  3 | 1/5  |               0.2
  2 | 3/7  | 0.428571428571429
  1 | 2/3  | 0.666666666666667
(3 rows)

-- cast
SELECT '1/3'::fraction::float8;
      float8       
-------------------
 0.333333333333333
(1 row)

SELECT 3::fraction;
 fraction 
----------
 3/1
(1 row)

